docs/proposals/drafts/20211007-pluggable-authenticators.md

```diff
@@ -0,0 +1,146 @@
+---
+toc: true
+title: 2021-10 Pluggable Authenticators
+---
+
+## Introduce Pluggable Authenticators
+
+* **Owners**
+  * @tareqmamari
+
+## TL; DR
+
+We propose a pluggable authenticator approach to support plugging in custom authenticators in a self-registering manner.
+
+## Why
+
+This proposed approach allows the adopters of Observatorium to use their own custom authenticators' implementation other than the current OIDC, mTLS or Openshift authenticators.
+
+## Goals
+
+- Provide a common interface for the authenticators, so that the adopters can use it to implement their own.
+- Provide a self-registering mechanism of the authenticators so that the adopters do not need to patch the current codebase to introduce their own authenticators.
+- Adopters should be able to configure the tenants, in a way it can specify which authenticator type to use for each tenant.
+- The new changes must be backward-compatible. In other words, the current authentication implementation and configuration must not be broken by this change.
+
+## How
+
+Currently, Observatorium API supports only OIDC and mTLS based authentication, however, there are adopters who do not use standard OIDC or mTLS, instead, they use for example jwt-based access tokens with special handling that is done using private go modules owned by those adopters.
+
+We propose to refactor the current way of initializing the authenticators to introduce a common interface, which means all new authenticators have the same basic functions where the adopters need to implement for their custom authenticators. In addition, in this proposal, we aim to provide a self-registering mechanism, so that introducing any new authenticator does not necessary require any additional patch/change in the existing codebase, instead, only the plugin's go source file is needed other than patching the other source files to add the plugin. For that, we propose to use importing for side-effect mechanism, where a protected map of authenticator type and their factories can be used for such scenario.
+
+## Implementation Options
+
+1. Importing for side-effect
+2. Go Plugin Module
+3. Oauth2 Providers
+
+## 1. Importing for side-effects
+
+Like [Option 3](#3-oauth2-providers) except that we do not have a static list of supported authenticators. Instead, each authenticator can register themselves through a map of authenticator types and their factories.
+
+To onboard a new authenticator, the following is needed:
+
+1. Implement the corresponding interface.
+2. Implement a factory.
+3. Register the factory in the authenticators' factories map, this is the part where it registers itself.
+
+The factory should receive three parameters: a config map of strings and interfaces, tenant's name, and a logger instance.
```

Comment:
I don't think we need to go into so much detail in the proposal. The parameters that a function receives should be part of the low level implementation rather than the high level proposal.